File: api/src/doctor.ts
import * as dotenv from 'dotenv';
import * as path from 'path';
import * as fs from 'fs';
import { createClient } from '@supabase/supabase-js';
import { MongoClient } from 'mongodb';
import { RoomServiceClient } from 'livekit-server-sdk';
import { Redis } from '@upstash/redis';
import ws from 'ws';

// Colors for formatting
const colors = {
  ok: '\x1b[32m[OK]\x1b[0m',
  warn: '\x1b[33m[WARN]\x1b[0m',
  fail: '\x1b[31m[FAIL]\x1b[0m',
  info: '\x1b[36m[INFO]\x1b[0m',
  bold: '\x1b[1m',
  reset: '\x1b[0m',
};

// Mask sensitive credentials in MongoDB URI
function maskMongoUri(uri: string): string {
  const match = uri.match(/^(mongodb(?:\+srv)?:\/\/)([^/\s?]+)/i);
  if (!match) return 'mongodb://***@unknown...';
  const scheme = match[1];
  const authority = match[2];
  
  const atIndex = authority.lastIndexOf('@');
  const host = atIndex !== -1 ? authority.substring(atIndex + 1) : authority;
  
  let maskedHost = host;
  const clusterMatch = host.match(/^(cluster)/i);
  if (clusterMatch) {
    maskedHost = clusterMatch[1] + '...';
  } else if (host.length > 10) {
    maskedHost = host.substring(0, 7) + '...';
  }
  
  return `${scheme}***@${maskedHost}`;
}

// Sanitize URL by removing credentials if any are present
function sanitizeUrl(url: string): string {
  const cleanUrl = url.replace(/^([a-zA-Z+.-]+:\/\/)(?:[^@]+)@(.*)$/i, '$1$2');
  const match = cleanUrl.match(/^([a-zA-Z+.-]+:\/\/)([^/]+)/i);
  if (!match) return url;
  
  const protocol = match[1];
  const host = match[2];
  
  if (host.toLowerCase().endsWith('.upstash.io')) {
    return `${protocol}***.upstash.io`;
  }
  
  const parts = host.split('.');
  if (parts.length > 2) {
    return `${protocol}***.${parts.slice(1).join('.')}`;
  }
  
  return `${protocol}***`;
}

// Mask sensitive environment variable values to prevent leakage in error outputs
function maskSecrets(text: string): string {
  let masked = text;
  masked = masked.replace(/(mongodb(?:\+srv)?:\/\/[^:]+:)([^@]+)(@)/ig, '$1***$3');
  for (const key of ['SUPABASE_SERVICE_ROLE_KEY', 'LIVEKIT_API_SECRET', 'UPSTASH_REDIS_REST_TOKEN', 'MONGO_URI']) {
    const secret = process.env[key];
    if (secret && secret.length > 4) {
      masked = masked.split(secret).join('***');
    }
  }
  return masked;
}

// Help / Usage
const verbose = process.argv.includes('--verbose');

// Load environment variables
const envPath = path.resolve(__dirname, '../.env');
if (fs.existsSync(envPath)) {
  dotenv.config({ path: envPath });
} else {
  dotenv.config(); // fallback
}

console.log(`${colors.info} ${colors.bold}Starting API Environment Diagnostics...${colors.reset}\n`);

// Config Validation Rules
const urlRegex = /^https?:\/\/[^\s$.?#].[^\s]*$/i;
const mongoRegex = /^mongodb(\+srv)?:\/\/.+/i;
const wsRegex = /^(wss?|https?):\/\/[^\s$.?#].[^\s]*$/i; // LiveKit URLs can be ws/wss/http/https

interface EnvVar {
  key: string;
  required: boolean;
  validate?: (val: string) => boolean;
  formatError?: string;
}

const expectedEnvVars: EnvVar[] = [
  {
    key: 'SUPABASE_URL',
    required: true,
    validate: (val) => urlRegex.test(val),
    formatError: 'Must be a valid HTTPS URL',
  },
  {
    key: 'SUPABASE_SERVICE_ROLE_KEY',
    required: true,
  },
  {
    key: 'MONGO_URI',
    required: true,
    validate: (val) => mongoRegex.test(val),
    formatError: 'Must be a valid MongoDB URI (starting with mongodb:// or mongodb+srv://)',
  },
  {
    key: 'LIVEKIT_URL',
    required: true,
    validate: (val) => wsRegex.test(val),
    formatError: 'Must be a valid ws/wss or http/https URL',
  },
  {
    key: 'LIVEKIT_API_KEY',
    required: true,
  },
  {
    key: 'LIVEKIT_API_SECRET',
    required: true,
  },
  {
    key: 'UPSTASH_REDIS_REST_URL',
    required: true,
    validate: (val) => urlRegex.test(val),
    formatError: 'Must be a valid HTTP/HTTPS URL',
  },
  {
    key: 'UPSTASH_REDIS_REST_TOKEN',
    required: true,
  },
];

let validationFailed = false;
const values: Record<string, string> = {};

console.log(`${colors.bold}--- Environment Variables Check ---${colors.reset}`);
for (const envVar of expectedEnvVars) {
  const val = process.env[envVar.key];
  if (!val) {
    console.log(`${colors.fail} ${envVar.key} is missing`);
    validationFailed = true;
    continue;
  }
  
  if (envVar.validate && !envVar.validate(val)) {
    console.log(`${colors.fail} ${envVar.key} has invalid format. ${envVar.formatError}`);
    validationFailed = true;
    continue;
  }

  values[envVar.key] = val;
  console.log(`${colors.ok} ${envVar.key} is set and format is valid`);
}

if (validationFailed) {
  console.log(`\n${colors.fail} ${colors.bold}Environment validation failed. Please fix your .env file before running diagnostics.${colors.reset}`);
  process.exit(1);
}

console.log(`\n${colors.bold}--- External Services Connectivity Checks ---${colors.reset}`);

let exitCode = 0;

function logOk(msg: string) {
  console.log(`${colors.ok} ${msg}`);
}

function logWarn(msg: string) {
  console.log(`${colors.warn} ${msg}`);
}

function logFail(msg: string, err?: any) {
  console.log(`${colors.fail} ${msg}`);
  if (err) {
    if (verbose) {
      const stack = err.stack || String(err);
      console.error(maskSecrets(stack));
    } else {
      console.error(`  Reason: ${maskSecrets(err.message || String(err))}`);
    }
  }
}

// Timeout wrapper helper
function withTimeout<T>(promise: Promise<T>, ms: number, errorMessage = 'Timeout exceeded'): Promise<T> {
  return Promise.race([
    promise,
    new Promise<T>((_, reject) => setTimeout(() => reject(new Error(errorMessage)), ms)),
  ]);
}

async function runDiagnostics() {
  // 1. Supabase Check
  try {
    // Set global.WebSocket so Supabase client does not warn/fail in Node.js < 22
    (global as any).WebSocket = ws;
    const supabase = createClient(values.SUPABASE_URL, values.SUPABASE_SERVICE_ROLE_KEY, {
      auth: { persistSession: false },
    });

    // Real client query to check scenarios table
    const supabaseCheck = supabase.from('scenarios').select('id').limit(1);
    const response = (await withTimeout(Promise.resolve(supabaseCheck), 8000, 'Connection to Supabase timed out after 8s')) as any;

    if (response.error) {
      // 42P01 is Postgres error for "relation does not exist" (meaning scenarios table does not exist)
      if (response.error.code === '42P01') {
        logWarn(`Supabase reachable and authenticated, but 'scenarios' table is missing. Did you run migrations?`);
      } else {
        throw response.error;
      }
    } else {
      logOk(`Supabase connected and 'scenarios' table exists`);
    }
  } catch (err: any) {
    exitCode = 1;
    logFail('Supabase connectivity or authentication failed', err);
  }

  // 2. MongoDB Check
  try {
    const maskedMongoUri = maskMongoUri(values.MONGO_URI);
    const mongoClient = new MongoClient(values.MONGO_URI, {
      serverSelectionTimeoutMS: 5000,
    });
    
    await withTimeout(mongoClient.connect(), 6000, 'MongoDB connection timed out after 6s');
    await mongoClient.db().admin().command({ ping: 1 });
    await mongoClient.close();
    logOk(`MongoDB connected successfully (using ${maskedMongoUri})`);
  } catch (err: any) {
    exitCode = 1;
    logFail('MongoDB connection failed', err);
  }

  // 3. LiveKit Check
  try {
    // Normalize ws/wss -> http/https for RoomServiceClient rest connection
    let normalizedLiveKitUrl = values.LIVEKIT_URL;
    if (normalizedLiveKitUrl.startsWith('wss://')) {
      normalizedLiveKitUrl = normalizedLiveKitUrl.replace('wss://', 'https://');
    } else if (normalizedLiveKitUrl.startsWith('ws://')) {
      normalizedLiveKitUrl = normalizedLiveKitUrl.replace('ws://', 'http://');
    }

    const roomService = new RoomServiceClient(
      normalizedLiveKitUrl,
      values.LIVEKIT_API_KEY,
      values.LIVEKIT_API_SECRET
    );
    
    await withTimeout(roomService.listRooms(), 8000, 'LiveKit room listing timed out after 8s');
    logOk('LiveKit authenticated successfully (room list fetched)');
  } catch (err: any) {
    exitCode = 1;
    logFail('LiveKit authentication or reachability failed', err);
  }

  // 4. Upstash Redis Check
  try {
    const sanitizedRedisUrl = sanitizeUrl(values.UPSTASH_REDIS_REST_URL);
    const redis = new Redis({
      url: values.UPSTASH_REDIS_REST_URL,
      token: values.UPSTASH_REDIS_REST_TOKEN,
    });

    const testKey = `doctor_healthcheck_${Date.now()}`;
    
    // Set
    await withTimeout(redis.set(testKey, 'ok', { ex: 10 }), 5000, 'Upstash Redis WRITE operation timed out');
    
    // Get
    const val = await withTimeout(redis.get(testKey), 5000, 'Upstash Redis READ operation timed out');
    if (val !== 'ok') {
      throw new Error(`Integrity error: Expected 'ok', but got '${val}'`);
    }

    // Delete
    await withTimeout(redis.del(testKey), 5000, 'Upstash Redis DELETE operation timed out');
    
    logOk(`Upstash Redis read/write/delete healthcheck succeeded (using ${sanitizedRedisUrl})`);
  } catch (err: any) {
    exitCode = 1;
    logFail('Upstash Redis healthcheck failed', err);
  }

  // Final status report
  console.log('\n-----------------------------------------');
  if (exitCode === 0) {
    console.log(`${colors.ok} ${colors.bold}API environment diagnostics passed successfully!${colors.reset}`);
  } else {
    console.log(`${colors.fail} ${colors.bold}API environment diagnostics failed. See errors above.${colors.reset}`);
  }
  process.exit(exitCode);
}

runDiagnostics();

File: agent/doctor.py
import core.bootstrap
import os
import sys
import re
import time
import urllib.request
import json
import asyncio
import ssl

# Terminal coloring
class Colors:
    OK = '\033[92m[OK]\033[0m'
    WARN = '\033[93m[WARN]\033[0m'
    FAIL = '\033[91m[FAIL]\033[0m'
    INFO = '\033[94m[INFO]\033[0m'
    BOLD = '\033[1m'
    RESET = '\033[0m'

# Check for verbose flag
verbose = "--verbose" in sys.argv

print(f"{Colors.INFO} {Colors.BOLD}Starting Agent Environment Diagnostics...{Colors.RESET}\n")

# Format patterns
url_regex = re.compile(r'^https?://[^\s$.?#].[^\s]*$', re.IGNORECASE)
mongo_regex = re.compile(r'^mongodb(\+srv)?://.+', re.IGNORECASE)
livekit_regex = re.compile(r'^(wss?|https?)://[^\s$.?#].[^\s]*$', re.IGNORECASE)

def mask_mongo_uri(uri: str) -> str:
    match = re.match(r'^(mongodb(?:\+srv)?://)([^/\s?]+)', uri, re.IGNORECASE)
    if not match:
        return 'mongodb://***@unknown...'
    scheme = match.group(1)
    authority = match.group(2)
    
    at_index = authority.rfind('@')
    host = authority[at_index + 1:] if at_index != -1 else authority
    
    cluster_match = re.match(r'^(cluster)', host, re.IGNORECASE)
    if cluster_match:
        masked_host = cluster_match.group(1) + '...'
    elif len(host) > 10:
        masked_host = host[:7] + '...'
    else:
        masked_host = host
        
    return f"{scheme}***@{masked_host}"

def sanitize_url(url: str) -> str:
    # First, remove any credentials (e.g. username:password@)
    clean_url = re.sub(r'^([a-zA-Z+.-]+://)(?:[^@]+)@(.*)$', r'\1\2', url, flags=re.IGNORECASE)
    
    # Match protocol and host
    match = re.match(r'^([a-zA-Z+.-]+://)([^/]+)', clean_url, re.IGNORECASE)
    if not match:
        return url
        
    protocol = match.group(1)
    host = match.group(2)
    
    if host.lower().endswith('.upstash.io'):
        return f"{protocol}***.upstash.io"
        
    parts = host.split('.')
    if len(parts) > 2:
        return f"{protocol}***.{'.'.join(parts[1:])}"
        
    return f"{protocol}***"

def mask_secrets(text: str) -> str:
    masked = text
    # Mask MongoDB URI credentials (e.g. mongodb+srv://user:pass@host)
    masked = re.sub(r'(mongodb(?:\+srv)?://[^:]+:)([^@]+)(@)', r'\1***\3', masked, flags=re.IGNORECASE)
    
    # Mask any direct mentions of the sensitive env var values
    sensitive_keys = [
        "GROQ_API_KEY", "DEEPGRAM_API_KEY", "LIVEKIT_API_SECRET", 
        "MONGO_URI", "UPSTASH_REDIS_REST_TOKEN", "SUPABASE_SERVICE_ROLE_KEY", "SUPABASE_ANON_KEY"
    ]
    for key in sensitive_keys:
        secret = os.getenv(key)
        if secret and len(secret) > 4:
            masked = masked.replace(secret, "***")
    return masked

def create_verified_ssl_context():
    try:
        import certifi
        return ssl.create_default_context(cafile=certifi.where())
    except Exception:
        return ssl.create_default_context()

def get_ca_file():
    try:
        import certifi
        return certifi.where()
    except Exception:
        return None

expected_vars = [
    {"key": "GROQ_API_KEY", "required": True},
    {"key": "DEEPGRAM_API_KEY", "required": True},
    {"key": "LIVEKIT_URL", "required": True, "regex": livekit_regex, "error": "Must be a valid ws/wss or http/https URL"},
    {"key": "LIVEKIT_API_KEY", "required": True},
    {"key": "LIVEKIT_API_SECRET", "required": True},
    {"key": "MONGO_URI", "required": True, "regex": mongo_regex, "error": "Must be a valid MongoDB URI (starting with mongodb:// or mongodb+srv://)"},
    {"key": "UPSTASH_REDIS_REST_URL", "required": True, "regex": url_regex, "error": "Must be a valid HTTP/HTTPS URL"},
    {"key": "UPSTASH_REDIS_REST_TOKEN", "required": True},
    {"key": "SUPABASE_URL", "required": True, "regex": url_regex, "error": "Must be a valid HTTP/HTTPS URL"},
    {"key": "SUPABASE_SERVICE_ROLE_KEY", "required": True},
    {"key": "SUPABASE_ANON_KEY", "required": True}
]

validation_failed = False
values = {}

print(f"{Colors.BOLD}--- Environment Variables Check ---{Colors.RESET}")
for var in expected_vars:
    key = var["key"]
    val = os.getenv(key)
    
    if not val:
        print(f"{Colors.FAIL} {key} is missing")
        validation_failed = True
        continue

    if key == "MONGO_URI" and val.startswith("MONGO_URI="):
        print(f"{Colors.FAIL} {key} appears to include the key name inside the value. Use `MONGO_URI=mongodb://...` or `MONGO_URI=mongodb+srv://...`")
        validation_failed = True
        continue
        
    if "regex" in var and not var["regex"].match(val):
        print(f"{Colors.FAIL} {key} has invalid format. {var['error']}")
        validation_failed = True
        continue
        
    values[key] = val
    print(f"{Colors.OK} {key} is set and format is valid")

if validation_failed:
    print(f"\n{Colors.FAIL} {Colors.BOLD}Environment validation failed. Please fix your .env file before running diagnostics.{Colors.RESET}")
    sys.exit(1)


print(f"\n{Colors.BOLD}--- External Services Connectivity Checks ---{Colors.RESET}")

exit_code = 0

def log_ok(msg):
    print(f"{Colors.OK} {msg}")

def log_warn(msg):
    print(f"{Colors.WARN} {msg}")

def log_fail(msg, err=None):
    print(f"{Colors.FAIL} {msg}")
    if err:
        if verbose:
            import traceback
            tb_str = traceback.format_exc()
            print(mask_secrets(tb_str))
        else:
            print(f"  Reason: {mask_secrets(str(err))}")

async def run_diagnostics():
    global exit_code
    
    # 1. MongoDB Check
    try:
        masked_mongo_uri = mask_mongo_uri(values["MONGO_URI"])
        from pymongo import MongoClient
        client = MongoClient(
            values["MONGO_URI"],
            serverSelectionTimeoutMS=5000,
            tls=True,
            tlsCAFile=get_ca_file(),
        )
        # Verify connectivity
        client.admin.command("ping")
        client.close()
        log_ok(f"MongoDB connected successfully (using {masked_mongo_uri})")
    except Exception as e:
        exit_code = 1
        log_fail("MongoDB connection failed", e)

    # 2. Upstash Redis Check
    try:
        sanitized_redis_url = sanitize_url(values["UPSTASH_REDIS_REST_URL"])
        from upstash_redis import Redis
        redis = Redis(url=values["UPSTASH_REDIS_REST_URL"], token=values["UPSTASH_REDIS_REST_TOKEN"])
        test_key = f"doctor_healthcheck_{int(time.time())}"
        
        # Read/Write/Delete healthcheck
        redis.set(test_key, "ok", ex=10)
        val = redis.get(test_key)
        if val != "ok":
            raise Exception(f"Integrity check failed: expected 'ok', got '{val}'")
        redis.delete(test_key)
        
        log_ok(f"Upstash Redis read/write/delete healthcheck succeeded (using {sanitized_redis_url})")
    except Exception as e:
        exit_code = 1
        log_fail("Upstash Redis healthcheck failed", e)

    # 3. LiveKit Check
    client = None
    session = None
    try:
        # Normalize ws/wss -> http/https for RoomServiceClient rest connection
        normalized_livekit_url = values["LIVEKIT_URL"]
        if normalized_livekit_url.startswith("wss://"):
            normalized_livekit_url = normalized_livekit_url.replace("wss://", "https://")
        elif normalized_livekit_url.startswith("ws://"):
            normalized_livekit_url = normalized_livekit_url.replace("ws://", "http://")

        from livekit import api
        import aiohttp
        ssl_context = create_verified_ssl_context()
        session = aiohttp.ClientSession(connector=aiohttp.TCPConnector(ssl=ssl_context))
        # LiveKitAPI is the container for services in the livekit-api package
        client = api.LiveKitAPI(
            normalized_livekit_url,
            values["LIVEKIT_API_KEY"],
            values["LIVEKIT_API_SECRET"],
            session=session
        )
        
        # list_rooms returns an awaitable
        await asyncio.wait_for(client.room.list_rooms(api.ListRoomsRequest()), timeout=8.0)
        log_ok("LiveKit authenticated successfully (room list fetched)")
    except Exception as e:
        exit_code = 1
        log_fail("LiveKit authentication or reachability failed", e)
    finally:
        if client is not None:
            await client.aclose()
        if session is not None:
            await session.close()

    # 4. Groq Check
    try:
        from groq import Groq
        client = Groq(api_key=values["GROQ_API_KEY"])
        # Validate authentication using lightweight models metadata call
        client.models.list()
        log_ok("Groq authenticated successfully (models list fetched)")
    except Exception as e:
        exit_code = 1
        log_fail("Groq authentication failed", e)

    # 5. Deepgram Check
    try:
        # Validate authentication using lightweight projects endpoint via urllib
        req = urllib.request.Request(
            "https://api.deepgram.com/v1/projects",
            headers={"Authorization": f"Token {values['DEEPGRAM_API_KEY']}"}
        )
        
        # Use certifi-backed CA roots when available to avoid local OpenSSL trust-store issues.
        context = create_verified_ssl_context()
        with urllib.request.urlopen(req, timeout=5.0, context=context) as response:
            res_data = json.loads(response.read().decode())
            if "projects" not in res_data:
                raise Exception("Response format invalid: expected 'projects' key")
                
        log_ok("Deepgram authenticated successfully (projects list fetched)")
    except Exception as e:
        exit_code = 1
        log_fail("Deepgram authentication failed", e)

    # 6. Supabase Check
    try:
        from supabase import create_client
        supabase = create_client(values["SUPABASE_URL"], values["SUPABASE_SERVICE_ROLE_KEY"])
        
        # Check scenarios table
        scenarios_res = supabase.table("scenarios").select("id").limit(1).execute()
        log_ok("Supabase connected and 'scenarios' table is accessible")
        
        # Check memories table
        memories_res = supabase.table("memories").select("id").limit(1).execute()
        log_ok("Supabase connected and 'memories' table is accessible")
    except Exception as e:
        exit_code = 1
        log_fail("Supabase connectivity or authentication failed", e)

    # Final summary report
    print("\n-----------------------------------------")
    if exit_code == 0:
        print(f"{Colors.OK} {Colors.BOLD}Agent environment diagnostics passed successfully!{Colors.reset if hasattr(Colors, 'reset') else Colors.RESET}")
    else:
        print(f"{Colors.FAIL} {Colors.BOLD}Agent environment diagnostics failed. See errors above.{Colors.reset if hasattr(Colors, 'reset') else Colors.RESET}")
        
    sys.exit(exit_code)

if __name__ == "__main__":
    asyncio.run(run_diagnostics())
